jacsstorage-services/src/main/java/org/janelia/jacsstorage/service/impl/FileSystemStorageService.java

```diff
@@ -0,0 +1,209 @@
+                Stream<Path> selectedFiles;
+                if (contentAccessParams.getEntriesCount() > 0) {
+                    selectedFiles = matchingFiles.skip(Math.max(contentAccessParams.getStartEntryIndex(), 0))
+                            .limit(contentAccessParams.getEntriesCount());
+                } else {
+                    selectedFiles = matchingFiles.skip(Math.max(contentAccessParams.getStartEntryIndex(), 0));
+                }
+                // if directories only returned nodes only have directories,
+                // otherwise they will have both files and directories
+                return selectedFiles
+                        .filter(p -> !contentAccessParams.isDirectoriesOnly() || Files.isDirectory(p))
+                        .map(this::createContentNode)
+                        .collect(Collectors.toList())
+                        ;
+            } catch (Exception e) {
+                throw new ContentException("Error reading directory content from: " + contentPath, e);
+            } finally {
+                LOG.info("List content {} with {} - {} secs", contentPath, contentAccessParams, (System.currentTimeMillis() - startTime) / 1000.);
+            }
+        } else if (Files.isRegularFile(contentPath, LinkOption.NOFOLLOW_LINKS)) {
+            return Collections.singletonList(createContentNode(contentPath));
+        } else {
+            throw new ContentException("Cannot handle reading content from " + contentPath);
+        }
+    }
+
+    private ContentNode createContentNode(Path p) {
+        try {
+            BasicFileAttributes fa = Files.readAttributes(p, BasicFileAttributes.class);
+            Path parent = p.getParent();
+            return new ContentNode(JacsStorageType.FILE_SYSTEM, JADEStorageURI.createStoragePathURI("", JADEOptions.create()))
+                    .setName(p.getFileName().toString())
+                    .setPrefix(parent != null ? parent.toString() : "")
+                    .setSize(Files.isDirectory(p) ? 0 : fa.size())
+                    .setLastModified(new Date(fa.lastModifiedTime().toMillis()))
+                    .setCollection(Files.isDirectory(p))
+                    ;
+        } catch (Exception e) {
+            throw new ContentException(e);
+        }
+    }
+
+    @Override
+    public InputStream getContentInputStream(String contentLocation) {
+        Path contentPath = Paths.get(contentLocation);
+
+        if (Files.exists(contentPath)) {
+            if (Files.isRegularFile(contentPath)) {
+                try {
+                    return Files.newInputStream(contentPath);
+                } catch (IOException e) {
+                    throw new ContentException(e);
+                }
+            } else {
+                throw new ContentException("Content found at " + contentLocation + " is not a regular file");
+            }
+        } else {
+            throw new NoContentFoundException("No object found at " + contentLocation);
```

Comment:
Here file not found is treated correctly, as an exception. Why doesn't listContentNodes use the same pattern?